### ISSUE TITLE
Inference of lambda passed as an argument to an overloaded function doesn't work

### ISSUE THREAD

**Issue by hamdanal** [astral-sh/ty#3597]
### Summary

In this example, inference of the lambda parameters and return type work with the non-overloaded function but fails with the overloaded one even though the overload is correctly resolved (as evidenced by the outer `reveal_type`s)

```py
from collections.abc import Callable
from typing import overload, reveal_type

@overload
def f(a: str, b: Callable[[int], str], /) -> str: ...
@overload
def f(a: bytes, b: Callable[[int], bytes], /) -> bytes: ...
def f(a, b, /): return a

def g[T: (str, bytes)](a: T, b: Callable[[int], T], /) -> T: return a

reveal_type(  # Revealed type: `str`
    f("", reveal_type(lambda x: str(x) * x))  # Revealed type: `(x) -> Unknown`
)
reveal_type(  # Revealed type: `bytes`
    f(b"", reveal_type(lambda x: x.to_bytes(1)))  # Revealed type: `(x) -> Unknown`
)
reveal_type(  # Revealed type: `str`
    g("", reveal_type(lambda x: str(x) * x))  # Revealed type: `(x: int) -> str`
)
reveal_type(  # Revealed type: `bytes`
    g(b"", reveal_type(lambda x: x.to_bytes(1)))  # Revealed type: `(x: int) -> bytes`
)
```

[Playground](https://play.ty.dev/2ac8a9c7-ff04-4494-ab00-bc9be8393825)

### Version

ty 0.0.40

**Comment by carljm** [comment]
Thanks for the report!

It looks like pyright does better here, but we share our current limitation with all other type checkers.

The root cause is that overloaded-call argument inference is speculative.

When checking an overloaded call, we first commit types from a context-free inference of the argument, then re-infer against each overload only for call checking: [builder.rs](https://github.com/astral-sh/ruff/blob/6b717e8d553a8076f6927894e221064a923aa13a/crates/ty_python_semantic/src/types/infer/builder.rs#L5572-L5621).

Lambda-body inference later recovers parameter types from the committed enclosing statement: [function.rs](https://github.com/astral-sh/ruff/blob/6b717e8d553a8076f6927894e221064a923aa13a/crates/ty_python_semantic/src/types/infer/builder/function.rs#L1149-L1171). It therefore sees `(x) -> Unknown` (from the context-free inference), not the better type from the speculative `Callable[[int], ...]` context.

That is why the outer overload resolves correctly while the inner lambda in [ty#3597](https://github.com/astral-sh/ty/issues/3597) remains unknown.

One fix might be to pass the active contextual lambda signature into lambda-scope inference directly.

#3001 and #3278 are related.
